### ISSUE TITLE
Improve parallel heap collection

### ISSUE THREAD

**Issue by YorickPeterse** [inko-lang/inko#64]
Currently the heap collector processes execution contexts (= call frames) in parallel. Based on various benchmarks I had when implementing the GC I found this yielded _much_ better performance compared to processing individual pointers in parallel. If I remember correctly in the current setup the collector can trace/mark about 1 million live objects in roughly 10 milliseconds, but this assumes ideal conditions (e.g. no evacuating, promoting, etc).

While most processes will have more than 1 execution context, it _is_ possible that only a few of these will contain pointers. In this case it's possible for only a small percentage of the available to be performing any work.

If possible the setup should be changed so that it meets the following requirements:

1. It provides better performance in all cases compared to the current setup (20% faster at minimum)
2. It does not require implementing complex synchronisation code (e.g. I prefer using an existing library, similar to how the collector currently uses rayon)

**Comment by YorickPeterse** [comment]
One way the GC could be optimized is to use atomic operations (CAS in particular) instead of locking when moving objects. For example, the bump allocator could work as follows:

1. Get the current bump pointer, store in local `a`
2. Add 32 bytes to `a`
3. Check if this value is within the allowed range
4. If so, swap the bump pointer with `a` using an atomic compare-and-swap
5. If we won, write data to the pointer; otherwise try again

This means we only need locking when adding a new block to a bucket.

For the source pointer we need to do something similar. For example:

1. Get the pointer we need to evacuate/promote
2. Get the prototype, store in local `a`
3. Set the prototype to `0x1` or some other reserved addresses (one that is not 0x0, and not used by the VM) using an atomic compare-and-swap
4. If we won, copy the object (using the above approach); otherwise skip the object
5. Set `a` as the prototype of the copy
6. Replace the old object's prototype with a forwarding pointer

In theory this _could_ work.

**Comment by YorickPeterse** [comment]
I'm thinking of the following approach to make things more parallel:

Instead of only processing stack frames in parallel we process blocks in parallel. To make this work we introduce a 3rd lazy allocated object map on a per block basis: the trace map. This map is allocated when starting a GC and free'd at the end, removing the 1 KB overhead outside of garbage collection cycles.

When we start a GC we'll have N threads scan the stack in parallel, marking objects in the trace map. The result is the roots as a list of entries in these maps. During this trace we don't mark the objects as live just yet.

Once this trace is done we divide the blocks across N threads. Each thread will scan over the trace maps of their blocks. For every marked entry we mark the object as live, then traverse any pointers stored in it. For every pointer that we find we mark it in the trace map. This process will repeat itself until a thread has no more entries in the trace maps of its blocks. 

Once all threads are done all trace maps are empty and we know which objects are live and which aren't. At this point we perform the usual GC steps, the trace map is free'd when we traverse all blocks to determine if they can be recycled, released, etc.

In this setup there are essentially two distinct traces:

1. A trace phase to determine and mark the roots
2. A trace phase to mark live objects and "schedule" new objects for marking

The benefit of this approach is that per block (= 1024 objects) we only need 1 KB of memory, instead of 8 KB plus Vec padding (if we were to store pointers in an array of sorts). We also don't need any synchronisation when scheduling new objects as object maps can be modified concurrently.

A downside of this approach is that in the above setup we can run into a case where 3 out of 4 threads are done while the 1 remaining thread is working its ass off marking objects. It's also possible that thread A may produce new entries in the trace map of thread B, requiring B to wake up again if it already finished its work. For this we could use a condition variable, but sending it a signal for every entry in the trace map will most likely add quite a lot of overhead. 

Another option is to have some kind of barrier in every thread that is reached once the thread has scanned over all blocks. If we reach this barrier we check if we marked any entries in the trace map, if so we re-process our blocks. If no threads marked any new entries they will all suspend. This will require some form of synchronisation to get a stable view, but this is only necessary once we have processed all blocks. Basically the most difficult part here is figuring out when we're done and when there's still remaining work to be done.
